### ISSUE TITLE
Has many through association: undefined method `klass' for nil:NilClass

### ISSUE THREAD

**Issue by xtagon** [thoughtbot/shoulda-matchers#611]
Hi!

I have an Subject model which should have many Photos through SubjectPhotoAssignments. When I test this, I get `undefined method 'klass' for nil:NilClass`. Here is my setup:

``` ruby
class Subject < ActiveRecord::Base
  belongs_to :project, touch: true
  has_many :photos, through: :subject_photo_assignments
end
```

``` ruby
class SubjectPhotoAssignment < ActiveRecord::Base
  belongs_to :photo
  belongs_to :subject, touch: true
end
```

``` ruby
class Photo < ActiveRecord::Base
  belongs_to :project, touch: true
  has_many :album_photo_assignments
  has_many :subject_photo_assignments
end
```

Here is the failing test:

``` ruby
RSpec.describe Subject, type: :model do
  it { should have_many(:photos).through(:subject_photo_assignments) }
end
```

```
  2) Subject should have many photos through subject_photo_assignments
     Failure/Error: it { should have_many(:photos).through(:subject_photo_assignments) }
     NoMethodError:
       undefined method `klass' for nil:NilClass
```

What am I doing wrong? Any help is greatly appreciated. Let me know if more context would be helpful.

I'm using v2.7.0

**Comment by 1v** [comment]
I have this error with code like this one:
```ruby
class Subject < ActiveRecord::Base
  has_one :subject_photo_assignment
  has_many :photos, through: :subject_photo_assignments
end
```

**Comment by mcmire** [comment]
@1v You have a `subject_photo_assignment` association, but you don't have a `subject_photo_assignments` association (note the "s" at the end).

**Comment by 1v** [comment]
@mcmire it's working with "s" too, but tests failing.

**Comment by mcmire** [comment]
@1v Okay. Have you tried:

``` ruby
class Subject < ActiveRecord::Base
  has_one :subject_photo_assignment
  has_many :photos, through: :subject_photo_assignment   # note that this is singular, not plural
end
```

**Comment by 1v** [comment]
@mcmire yeah that's how I do right now. But Rails accepts plural too. It took me time to figure out because error message not informative.

**Comment by mcmire** [comment]
@1v Okay, fair enough. I'll re-open this issue.

**Comment by esbanarango** [comment]
I also have this error with this "setup":

````ruby
module Hanuman
  class Observation < ActiveRecord::Base
    belongs_to :survey, touch: true
  end
end
````

````ruby
Hanuman::Observation.class_eval do
  has_one  :user, through: :survey
end
````

````ruby
module Hanuman
  RSpec.describe ObservationAnswer, type: :model do
    it { is_expected.to have_one(:user).through(:survey) }
  end
end
````


```
Failure/Error: it { is_expected.to have_one(:user).through(:survey) }
NoMethodError:
  undefined method `klass' for nil:NilClass
```

**Comment by mcmire** [comment]
Related issue: #646. #723 may solve both of these at once, I'll have to check.

**Comment by KapilSachdev** [comment]
I can verify that the reported issue has been fixed in the version `4.4.1`, so we can close this issue for now.

Before:
```ruby
     Failure/Error: it { should have_many(:words).through(:pag) }
     
     NoMethodError:
       undefined method `class_name' for nil:NilClass
       Did you mean?  class_eval
```
After:
```ruby
     Failure/Error: it { should have_many(:words).through(:pag) }
       Expected Book to have a has_many association called words through pages (Could not find the source association(s) "word" or :words in model Page. Try 'has_many :words, :through => :pages, :source => <name>'. Is it one of book?)
```
_Error text differs as i'm using rails `6.1.0.alpha`_

**Comment by mcmire** [comment]
Thanks! Closing.
